src/xrpld/app/misc/CredentialHelpers.cpp

```diff
@@ -225,6 +226,50 @@ makeSorted(STArray const& in)
     return out;
 }
 
+NotTEC
+checkArray(STArray const& credentials, unsigned maxSize, beast::Journal j)
+{
+    if (credentials.empty() || (credentials.size() > maxSize))
+    {
+        JLOG(j.trace()) << "Malformed transaction: "
+                           "Invalid credentials size: "
+                        << credentials.size();
+        return temMALFORMED;
+    }
+
+    std::unordered_set<uint256> duplicates;
+    for (auto const& credential : credentials)
+    {
+        auto const& issuer = credential[sfIssuer];
+        if (!issuer)
+        {
+            JLOG(j.trace()) << "Malformed transaction: "
+                               "Issuer account is invalid: "
+                            << to_string(issuer);
+            return temINVALID_ACCOUNT_ID;
+        }
+
+        auto const ct = credential[sfCredentialType];
+        if (ct.empty() || (ct.size() > maxCredentialTypeLength))
+        {
+            JLOG(j.trace()) << "Malformed transaction: "
+                               "Invalid credentialType size: "
+                            << ct.size();
+            return temMALFORMED;
+        }
+
+        auto [it, ins] = duplicates.insert(sha512Half(issuer, ct));
```

Comment:
nit: As far as efficiency of the `PDSet` transaction is concerned, is it necessary to do the `SHA512Half` operation for every input `Credential` ? Since there are at most 10 elements, I was wondering if it would be cheaper to use a `std::set` container.

I don't feel strongly about it, since there are only 10 elements anyway.

Reply:
Generally unordered containers are faster than ordered. May be there are some special cases but this part of the code doesn't worth to make additional researches.